pkg/webhook/instance_admission.go

```diff
@@ -0,0 +1,240 @@
+|------------------|-----------------|---------|--------------|------------|-------|---------------------------------------------------------------------------------------------------------------------|
+| x                | x               |         |              |            | [No]  | Forbid parameter updates when a plan is scheduled unless the same plan is triggered (instance status will be reset) |
+| x                |                 | x       |              |            | No    | Forbid upgrades if another plan is running                                                                          |
+| x                |                 |         | x            |            | No    | Forbid plan overrides (for now)                                                                                     |
+| x                |                 |         |              | x          | No    | Forbid plan cancellations (for now)                                                                                 |
+| ---              |                 |         |              |            |       | ---                                                                                                                 |
+|                  | x               |         | x            |            | No    | Forbid simultaneous parameter update and directly triggered plan                                                    |
+|                  |                 | x       | x            |            | No    | Forbid simultaneous upgrades and directly triggered plans                                                           |
+|                  | x               | x       |              |            | No*   | Forbid simultaneous upgrades and parameter updates                                                                  |
+ --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+ *Note: simultaneous upgrade and parameter update is NOT allowed. However, there is a exception where new OV removes an existing
+ parameter. Removing this parameter in the instance update would not count as parameter update (since there is no plan to trigger).
+
+ For the complete set of rules, see the corresponding test.
+*/
+
+// validateUpdate takes in the old and new (updated) instance and returns a new plan that might
+// be triggered based on the update and an error if the update is not valid. Return plan might be
+// - <nil> when there is no change to an existing scheduled plan
+// - '' empty string when an existing plan should be canceled (not implemented yet)
+// - 'newPlan' some new plan that should be triggered
+func validateUpdate(old, new *kudov1beta1.Instance, ov *kudov1beta1.OperatorVersion) (*string, error) {
```

Comment:
I just realized few things about this PR while working on something in controller-runtime :)

- this should not be called validate as we're calling reset status in between
- also if you want to reset status, you need to transform the webhook from validating to mutating in the webhook config

perhaps this PR should go in without the reset and the rest should be in another PR?

Reply:
I don't mind, `resetInstanceStatus` still returns `nil` here.